keps/sig-apps/2255-pod-cost/README.md

```diff
@@ -55,191 +49,210 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
+On scale down, the application may want to remove pods running on the most expensive 
+nodes first. For example, remove pods from nodes running on standard VMs first
+then from ones running on preemptible/spot VMs (which can be 80% cheaper than standard VMs).
 
 
 ### Risks and Mitigations
 
-On previous Kubernetes ReplicaSet controller versions that don't implement the pod-cost annotation feature, the same application might make false assumptions about the protection of a master instance or workers with open (user-)sessions on it. As the pod-cost annotation would be only a suggestion to the ReplicaSet controller, the application developer should, however, handle the case of a failed master instance or broken user sessions. The feature is just an improvement, not a guarantee, as there might happen timing issues between setting the annotation and the next controller scale-down event.
+One risk is that users perceive the feature as a guarantee to delete order. Documentation
+should stress the fact that this is best effort.
 
 ## Design Details
 
+The pod-delete-cost range will be from [-MaxInt, MaxInt]. The default value is 0.
+Invalid values (like setting the annotation to string) will be converted to
+the default value.
+
+When scaling down a ReplicaSet, controller-manager will prioritize deleting
+pods with lower `pod-delete-cost`. Specifically, the pod-delete-cost will be evaluated after
+step 3 and before step 4 as they are currently defined in
+[ActivePodsWithRanks](https://github.com/kubernetes/kubernetes/blob/cac933934b1301665e6e51a81c66c483f4e16c49/pkg/controller/controller_utils.go#L784-L809),
+which means the followig criteria is applied when comparing two pods regardless of their pod-delete-cost:
+- if one is assigned a node and the other is not, then the unassigned pod is deleted first.
+- if the two pods are in different phases, then the pod in pending/unknown status is deleted first.
+- if the two pods have different readiness status, then the not ready pod is deleted first
+
+
+If none of the pods set the pod-delete-cost annotation or all of them have the same value, then the 
+scale down behavior is not changed compared to now.
 
 ### Test Plan
 
 * Units test in kube-controller-manager package to test a variety of scenarios.
-* New E2E Tests to validate that replicas get deleted as expected e.g:
- * Replicas with lower pod-cost before replicas with higher pod-cost
- * Replicas with no pod-cost annotation set before replicas with low priority
+* Integration tests to validate that:
+ * Replicas with lower pod-delete-cost are deleted before replicas with higher pod-delete-cost
```

Comment:
More indentation of the tests? As printed in the review, `Integration tests ...`, `Replicas with lower ...` and `No behavior change ...` on the same level.

Reply:
done, thanks.